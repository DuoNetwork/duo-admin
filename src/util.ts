import moment from 'moment';
import request from 'request';
import * as CST from './constants';
import { IOption } from './types';

class Util {
	public log(text: any): void {
		console.log(moment().format('HH:mm:ss.SSS') + ' ' + text);
	}

	// public truncateNum(num: number, decimal: number = 3): number {
	// 	let x = num.toString();
	// 	x = x.slice(0, x.indexOf('.') + decimal);
	// 	return Number(x);
	// }

	public get(url: string, headerOthers?: object): Promise<any> {
		return new Promise((resolve, reject) =>
			request(
				{
					url,
					headers: Object.assign(
						{
							'user-agent': 'node.js'
						},
						headerOthers || {}

					)
				},
				(error, res, body) => {
					if (error) reject(error);
					else if (res.statusCode === 200) resolve(body);
					else reject('Error status ' + res.statusCode + ' ' + res.statusMessage);
				}
			)
		);
	}

	public generateRandomIdx(max: number, alpha: number): number[] {
		const output: number[] = [];
		if (alpha === 1) for (let i = 0; i < max; i++) output.push(i);
		else {
			const num = Math.floor(max * alpha);

			while (output.length < num) {
				const item = Math.floor(Math.random() * max);
				if (output.indexOf(item) < 0) output.push(item);
			}
		}

		return output;
	}

	public postJson(url: string, json: object): Promise<object> {
		return new Promise((resolve, reject) =>
			request(
				{
					url,
					method: 'POST',
					headers: {
						'Content-Type': 'application/json'
					},
					json
				},
				(error, res, body) => {
					if (error) reject(error);
					else if (res.statusCode === 200) resolve(body);
					else reject('Error status ' + res.statusCode + ' ' + res.statusMessage);
				}
			)
		);
	}

	public parseOptions(argv: string[]): IOption {
		const option = {
			live: process.argv.includes('live'),
			dynamo: process.argv.includes('dynamo'),
			aws: process.argv.includes('aws'),
			gcp: process.argv.includes('gcp'),
			azure: process.argv.includes('azure'),
			force: process.argv.includes('force'),
			gasPrice: 5e9,
			gasLimit: 200000,
			eth: 0,
			address: '',
			privateKey: '',
			price: 0,
			source: '',
			pwd: '',
			event: '',
			provider: '',
			contractState: '',
			accountNum: 1,
			saveAccount: false,
			from: '',
			to: '',
			value: 0,
			index: 0,
			total: 10,
			minEther: 0.02,
			alpha: 0.3,
			amtA: 0,
			amtB: 0,
			generator: ''
		};
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'gasPrice':
					option.gasPrice = Number(args[1]) || option.gasPrice;
					break;
				case 'gasLimit':
					option.gasLimit = Number(args[1]) || option.gasLimit;
					break;
				case 'eth':
					option.eth = Number(args[1]) || option.eth;
					break;
				case 'address':
					option.address = args[1] || option.address;
					break;
				case 'privateKey':
					option.privateKey = args[1] || option.privateKey;
					break;
				case 'price':
					option.price = Number(args[1]) || option.price;
					break;
				case 'pwd':
					option.pwd = args[1] || option.pwd;
					break;
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'event':
					option.event = args[1] || option.event;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				case 'contractState':
					option.contractState = args[1] || option.contractState;
					break;
				case 'accountNum':
					option.accountNum = Number(args[1]) || option.accountNum;
					break;
				case 'saveAccount':
					option.saveAccount = args[1] === 'yes' || option.saveAccount;
					break;
				case 'from':
					option.from = args[1] || option.from;
					break;
				case 'total':
					option.total = Number(args[1]) || option.total;
					break;
				case 'minEther':
					option.minEther = Number(args[1]) || option.minEther;
					break;
				case 'alpha':
					option.alpha = Number(args[1]) || option.alpha;
					break;
				case 'amtA':
					option.amtA = Number(args[1]) || option.amtA;
					break;
				case 'amtB':
					option.amtB = Number(args[1]) || option.amtB;
					break;
				case 'value':
					option.value = Number(args[1]) || option.value;
					break;
				case 'index':
					option.index = Number(args[1]) || option.index;
					break;
				case 'generator':
					option.generator = args[1] || option.generator;
					break;
				default:
					break;
			}
		}

		if (!option.provider)
			if (option.source === CST.SRC_MYETHER)
				option.provider = option.live
					? CST.PROVIDER_MYETHER_LIVE
					: CST.PROVIDER_MYETHER_DEV;
			else if (option.source === CST.SRC_INFURA)
				option.provider = option.live ? CST.PROVIDER_INFURA_LIVE : CST.PROVIDER_INFURA_DEV;
			else {
				option.provider = CST.PROVIDER_LOCAL_WS;
				option.source = '';
			}

		return option;
	}

	public getDynamoRole(tool: string, useDynamo: boolean): string {
		switch (tool) {
			case 'bitfinex':
			case 'gemini':
			case 'kraken':
			case 'gdax':
				return useDynamo ? CST.AWS_DYNAMO_ROLE_TRADE : CST.AWS_DYNAMO_ROLE_STATUS;
			case 'commit':
			case 'node':
				return CST.AWS_DYNAMO_ROLE_STATUS;
			case 'subscribe':
				return CST.AWS_DYNAMO_ROLE_EVENT;
			case 'hourly':
				return CST.AWS_DYNAMO_ROLE_HOURLY;
			case 'minutely':
				return CST.AWS_DYNAMO_ROLE_MINUTELY;
			default:
				return '';
		}
	}

	public getStatusProcess(tool: string, option: IOption) {
		let type = '';
		const platform = option.azure ? '_AZURE' : option.gcp ? '_GCP' : '_AWS';
		const privacy = option.dynamo ? '_PUBLIC' : '_PRIVATE';
		let source = '';

		switch (tool) {
			case 'bitfinex':
			case 'gemini':
			case 'kraken':
			case 'gdax':
				type = 'PRICE';
				source = '_' + tool.toUpperCase();
				break;
			case 'subscribe':
				type = 'EVENT';
				source =
					'_' +
					([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)
						? option.event.toUpperCase()
						: 'OTHERS');
				break;
			case 'commit':
				type = 'FEED';
				break;
			case 'hourly':
				type = 'HOURLY';
				break;
			case 'minutely':
				type = 'MINUTELY';
				break;
			case 'node':
				type = 'CHAIN';
				break;
			default:
				return '';
		}

		return type + platform + privacy + source;
	}

	public getNowTimestamp() {
		return moment().valueOf();
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}
}

const util = new Util();
export default util;
