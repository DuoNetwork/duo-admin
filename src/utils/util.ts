import moment from 'moment';
import * as CST from '../common/constants';
import { IOption } from '../common/types';
// import { Options } from 'aws-cli-js';

class Util {
	public logLevel: string = CST.LOG_INFO;

	private log(text: any, level?: string): void {
		if (level && CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	public composeQuery(paras: { [key: string]: any }): string {
		let query: string = '?';
		for (const key in paras) query += key + '=' + paras[key] + '&';
		query = query.slice(0, -1);
		return query;
	}

	public parseOptions(argv: string[]): IOption {
		const option = {
			forceREST: argv.includes('rest'),
			live: process.argv.includes('live'),
			dbLive: process.argv.includes('dbLive'),
			server: process.argv.includes('server'),
			dynamo: process.argv.includes('dynamo'),
			aws: process.argv.includes('aws'),
			gcp: process.argv.includes('gcp'),
			azure: process.argv.includes('azure'),
			force: process.argv.includes('force'),
			pair: '',
			contractType: 'Beethoven',
			tenor: 'Perpetual',
			assets: [''],
			sources: [''],
			exSources: [''],
			gasPrice: 5e9,
			gasLimit: 200000,
			source: '',
			event: '',
			provider: '',
			period: 1,
			base: 'USD',
			quote: 'ETH'
		};
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'exSources':
					option.exSources = args[1].split(',');
					break;
				case 'sources':
					option.sources = args[1].split(',');
					break;
				case 'pair':
					option.pair = args[1].replace('_', '|') || option.pair;
					break;
				case 'assets':
					option.assets = args[1].split(',');
					break;
				case 'gasPrice':
					option.gasPrice = Number(args[1]) || option.gasPrice;
					break;
				case 'gasLimit':
					option.gasLimit = Number(args[1]) || option.gasLimit;
					break;
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'event':
					option.event = args[1] || option.event;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				case 'period':
					option.period = Number(args[1]);
					break;
				case 'base':
					option.base = args[1] || option.base;
					break;
				case 'quote':
					option.quote = args[1] || option.quote;
					break;
				case 'contractType':
					option.contractType = args[1] || option.contractType;
					break;
				case 'tenor':
					option.tenor = args[1] || option.tenor;
					break;
				default:
					break;
			}
		}

		return option;
	}

	public getPeriodStartTimestamp(timestamp: number, period: number = 1) {
		// const now = this.getUTCNowTimestamp();
		return Math.floor(timestamp / 60000 / period - 1) * 60000 * period;
	}

	public timestampToString(ts: number) {
		//just for debugging purpose
		return moment.utc(ts).format('YYYY-MM-DDTHH:mm:ss');
	}

	public getStatusProcess(tool: string, option: IOption) {
		let type = '';
		const platform = option.azure ? '_AZURE' : option.gcp ? '_GCP' : '_AWS';
		console.log('dynamo', option.dynamo);
		const privacy = option.dynamo ? '_PUBLIC' : '_PRIVATE';
		let source = '';

		switch (tool) {
			case CST.TRADES:
				type = 'TRADE';
				source = option.source ? '_' + option.source.toUpperCase() : '';
				break;
			case CST.TRIGGER:
			case CST.FETCH_EVENTS:
				type = 'EVENT';
				source =
					'_' +
					([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)
						? option.event.toUpperCase()
						: 'OTHERS');
				break;
			case CST.COMMIT:
				type = 'FEED';
				break;
			case CST.CLEAN_DB:
				type = 'CLEANDB';
				break;
			case CST.DB_PRICES:
				type = 'PRICE';
				if (option.period === 1) source = '_MINUTELY';
				else if (option.period === 60) source = '_HOURLY';
				else source = '';
				break;
			case 'minutely':
				type = 'MINUTELY';
				break;
			case CST.FETCH_PRICE:
				type = 'FETCH';
				break;
			default:
				return '';
		}

		return type + platform + privacy + source;
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public async sleep(ms: number) {
		return new Promise(resolve => {
			setTimeout(resolve, ms);
		});
	}
}

const util = new Util();
export default util;
