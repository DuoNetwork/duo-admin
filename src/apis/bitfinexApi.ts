import { Constants, ITrade } from '@finbook/duo-market-data';
import * as CST from '../common/constants';
import {IBitfinexRawTradeRest, IBitfinexRawTradeWS } from '../common/types';
import httpUtil from '../utils/httpUtil';
import util from '../utils/util';
import BaseApi from './BaseApi';

const BFX = require('bitfinex-api-node');

export class BitfinexApi extends BaseApi {
	public source: string = Constants.API_BITFINEX;

	protected parseTradeWS(sourcePair: string, trade: IBitfinexRawTradeWS): ITrade {
		const { base, quote } = this.parseMarketData(sourcePair);
		return {
			base: base,
			quote: quote,
			source: this.source,
			id: trade.id + '',
			price: Number(trade.price),
			amount: Math.abs(Number(trade.amount)),
			timestamp: Number(trade.timestamp) * 1000
		};
	}

	protected parseTradeREST(sourcePair: string, trade: IBitfinexRawTradeRest): ITrade {
		const { base, quote } = this.parseMarketData(sourcePair);
		return {
			base: base,
			quote: quote,
			source: this.source,
			id: trade.tid + '',
			price: Number(trade.price),
			amount: Number(trade.amount),
			timestamp: trade.timestamp * 1000
		};
	}

	public async handleWSTradeMessage(m: string, sourcePair: string) {
		const data = JSON.parse(m);

		// Will ignore batch of trades, because snapshot does not contain trade id.
		// if single trade and is 'tu' event
		if (!Array.isArray(data) && data.hasOwnProperty('id') && sourcePair)
			await this.addTrades(
				this.sourcePairMapping[sourcePair],
				[this.parseTradeWS(sourcePair, data)],
				false
			);
	}

	public async fetchTradesREST(sourcePair: string) {
		const localPair = this.sourcePairMapping[sourcePair];
		const url: string =
			CST.API_BFX_BASE_URL +
			CST.API_BFX_VERSION +
			CST.API_BFX_TRADE +
			sourcePair +
			(this.last[localPair] ? util.composeQuery({ since: this.last[localPair] }) : '');
		util.logInfo(url);

		const result: IBitfinexRawTradeRest[] = JSON.parse(await httpUtil.get(url));
		await this.addTrades(
			this.sourcePairMapping[sourcePair],
			result.map(trade => this.parseTradeREST(sourcePair, trade))
		);
	}

	public fetchTradesWS(sourcePairs: string[]): any {
		const bfx = new BFX(null, null, {
			version: 1,
			transform: true
		});

		const w = bfx.ws;
		w.on('open', () => {
			util.logInfo('Connected');
			for (const pair of sourcePairs) {
				w.subscribeTrades(pair);
				util.logInfo(`Subscribed to trades ${pair}`);
			}
		});

		w.on('trade', (symbol: string, trades: any) => {
			this.handleWSTradeMessage(JSON.stringify(trades), symbol);
		});

		w.on('close', (code: number, reason: string) => {
			util.logError('connection closed ' + code + ' ' + reason);
			w.close();
			global.setTimeout(() => this.fetchTradesWS(sourcePairs), 1000);
		});

		w.on('error', (error: Error) => {
			util.logError(error);
			w.close();
			global.setTimeout(() => this.fetchTradesWS(sourcePairs), 1000);
		});
		return w;
	}
}
const bitfinexApi = new BitfinexApi();
export default bitfinexApi;
