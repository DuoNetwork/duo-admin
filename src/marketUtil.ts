import child_process from 'child_process';
import apis from './apis';
import { IOption } from './types';
import util from './util';

class MarketUtil {
	public async startFetching(tool: string, option: IOption): Promise<void> {
		const assetIds = option.assets.length
			? option.assets
			: option.pair.split('|').filter(asset => asset !== '');
		console.log('source' + option.source);

		if (!option.source) {
			util.logInfo('launching all sources for assets ' + assetIds.join(',') + ' in WS');
			for (const source in apis) {
				await util.sleep(1000);
				this.launchSource(tool, source, assetIds, option);
			}
		} else {
			const api = apis[option.source];
			if (!api) throw new Error('invalid source');
			api.init();

			const sourceInstruments = api.getSourceInstruments(assetIds);
			util.logInfo(
				`[${option.source}]:` +
					'start fetching for pairs ' +
					JSON.stringify(
						sourceInstruments.map(instrument => api.sourcePairMapping[instrument])
					)
			);
			apis[option.source].fetchTrades(sourceInstruments);
		}
	}

	public launchSource(tool: string, source: string, assets: string[], option: IOption) {
		util.logInfo(source);
		const cmd =
			`npm run ${tool} source=${source} assets=${assets.join(',')} ${
				option.forceREST ? 'rest' : ''
			}` +
			(process.platform === 'win32' ? '>>' : '&>') +
			` ${tool}.${source}.${assets.join('.')}.log`;

		util.logInfo(`[${source}]: ${cmd}`);

		const procInstance = child_process.exec(
			cmd,
			process.platform === 'win32' ? {} : { shell: '/bin/bash' }
		);

		if (!procInstance) util.logError('failed to launch ' + source);
		else util.logInfo(`[${source}]: Launched ${tool}  ${assets.join(',')}`);
	}
}

const marketUtil = new MarketUtil();
export default marketUtil;
