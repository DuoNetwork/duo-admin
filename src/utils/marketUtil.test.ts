import child_process from 'child_process';
import bitfinexApi from '../apis/bitfinexApi';
import marketUtil from './marketUtil';
import util from './util';

test('launchSource', () => {
	child_process.exec = jest.fn() as any;
	marketUtil.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
});

test('launchSource forceREST', () => {
	child_process.exec = jest.fn() as any;
	marketUtil.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: true,
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
});

test('launchSource debug', () => {
	child_process.exec = jest.fn() as any;
	marketUtil.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: true,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
});

test('launchSource live', () => {
	child_process.exec = jest.fn() as any;
	marketUtil.launchSource('tool', 'source', ['asset1', 'asset2', 'asset3'], {
		forceREST: false,
		debug: false,
		live: true
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
});

test('startFetching no source', async () => {
	util.sleep = jest.fn();
	marketUtil.launchSource = jest.fn();

	await marketUtil.startFetching('tool', {
		source: '',
		sources: [],
		exSources: [],
		assets: ['quote', 'base'],
		forceREST: false,
		debug: false
	} as any);

	await marketUtil.startFetching('tool', {
		source: '',
		sources: ['bitfinex', 'okex'],
		exSources: [],
		assets: ['quote', 'base'],
		forceREST: false,
		debug: false
	} as any);

	await marketUtil.startFetching('tool', {
		source: '',
		sources: ['bitfinex', 'okex'],
		exSources: [],
		pair: 'someQuote|someBase',
		assets: [],
		forceREST: false,
		debug: false
	} as any);

	expect((marketUtil.launchSource as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
});

test('startFetching source', async () => {
	util.sleep = jest.fn();
	bitfinexApi.getSourceInstruments = jest.fn(() => ['quote|base']);
	bitfinexApi.fetchTrades = jest.fn();
	await marketUtil.startFetching('tool', {
		source: 'bitfinex',
		assets: ['quote', 'base'],
		forceREST: true,
		debug: false
	} as any);
	expect((bitfinexApi.fetchTrades as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
});
