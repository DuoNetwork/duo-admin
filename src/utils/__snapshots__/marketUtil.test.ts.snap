// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`launchSource 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 >> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`launchSource debug 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 >> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`launchSource forceREST 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 rest>> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`launchSource live 1`] = `
Array [
  Array [
    "npm run tool source=source assets=asset1,asset2,asset3 >> tool.source.asset1.asset2.asset3.log",
    Object {},
  ],
]
`;

exports[`startFetching no source 1`] = `
Array [
  Array [
    "tool",
    "bitfinex",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [],
    },
  ],
  Array [
    "tool",
    "gemini",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [],
    },
  ],
  Array [
    "tool",
    "kraken",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [],
    },
  ],
  Array [
    "tool",
    "gdax",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [],
    },
  ],
  Array [
    "tool",
    "bitfinex",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [
        "bitfinex",
        "okex",
      ],
    },
  ],
  Array [
    "tool",
    "gemini",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [
        "bitfinex",
        "okex",
      ],
    },
  ],
  Array [
    "tool",
    "kraken",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [
        "bitfinex",
        "okex",
      ],
    },
  ],
  Array [
    "tool",
    "gdax",
    Array [
      "quote",
      "base",
    ],
    Object {
      "assets": Array [
        "quote",
        "base",
      ],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "source": "",
      "sources": Array [
        "bitfinex",
        "okex",
      ],
    },
  ],
  Array [
    "tool",
    "bitfinex",
    Array [
      "someQuote",
      "someBase",
    ],
    Object {
      "assets": Array [],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "pair": "someQuote|someBase",
      "source": "",
      "sources": Array [
        "bitfinex",
        "okex",
      ],
    },
  ],
  Array [
    "tool",
    "gemini",
    Array [
      "someQuote",
      "someBase",
    ],
    Object {
      "assets": Array [],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "pair": "someQuote|someBase",
      "source": "",
      "sources": Array [
        "bitfinex",
        "okex",
      ],
    },
  ],
  Array [
    "tool",
    "kraken",
    Array [
      "someQuote",
      "someBase",
    ],
    Object {
      "assets": Array [],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "pair": "someQuote|someBase",
      "source": "",
      "sources": Array [
        "bitfinex",
        "okex",
      ],
    },
  ],
  Array [
    "tool",
    "gdax",
    Array [
      "someQuote",
      "someBase",
    ],
    Object {
      "assets": Array [],
      "debug": false,
      "exSources": Array [],
      "forceREST": false,
      "pair": "someQuote|someBase",
      "source": "",
      "sources": Array [
        "bitfinex",
        "okex",
      ],
    },
  ],
]
`;

exports[`startFetching source 1`] = `
Array [
  Array [
    Array [
      "quote|base",
    ],
  ],
]
`;
